### ISSUE TITLE
Stabilizing Learning

### ISSUE THREAD

**Issue by chanb** [stanford-iprl-lab/GRAC#1]
Hi, thank you for writing this paper and open sourcing this code!
I really like the part where we can "get rid of the target network".

I'm trying to reproduce the paper's result with my own implementation and I have few questions
1. Learning is often stable but in one or two seeds that I am using, the return plummets to 0 (pretty much breaking the Q-function from what I am seeing). My implementation currently does not have any scheduling for alpha and learning rate. From your experience, how important is that?

2. Looking at the hyperparameters table from the paper, The CEM loss weight is somewhat arbitrary. Can you please give me some insight on these values? I also noticed that the actor is slightly smaller in learning rate, does 3e-4 hinder the agent from training well?

Thanks again and this paper is very interesting :)

**Comment by linsats** [comment]
Hi,  thank you for your interest in our work!

A1. Yes. The alpha and K is important for stabilizing the Q function. We report in our paper that " In practice, without appropriate K and alpha values, divergence usually happens within the first 50k training steps, thus it is quick to select appropriate values for K and alpha." If divergence happens, we would suggest to increase the values of K and alpha. Can you post more information about the experiments where the returns plummet to 0? Would you mind to share your implementation with us?

A2. The CEM loss weight is only hyperparameter to balance the two actor losses. We did not find they are sensitive to the performance.

**Comment by chanb** [comment]
Thank you for the response! This is the implementation: https://github.com/chanb/rl_sandbox_public/blob/master/rl_sandbox/algorithms/grac/grac.py

At ~700k steps, you can see that the `task0/return` (the training return) plummets to 0
![image](https://user-images.githubusercontent.com/16161287/100487785-72154d80-30d8-11eb-921f-a85df2fff591.png)

You can also see that there is a sudden change in a lot of these values
![image](https://user-images.githubusercontent.com/16161287/100487812-91ac7600-30d8-11eb-944c-87fb0f1cd9f6.png)

I already made one change back then where I noticed that I didn't clip the action before feeding to the Q-function during learning the Q-function (think the action was OOD in that case). Another potential issue is that my CEM has population size of 25. On paper it says it's 256 (if I read it right) but the source code seemed to be using way smaller population size. The issue seems to be gone after I crank that value up to 128 with the same seed. Will let you know if I see similar problem.

**Comment by linsats** [comment]
Thank you for sharing your implementation!

In our grac, the actor learning rate is adaptively changed based on the loss of the critic. https://github.com/stanford-iprl-lab/GRAC/blob/56eb20ebbda728e512f5bf9599f3417787f2da88/GRAC.py#L260

**Comment by chanb** [comment]
Wow I see. Totally missed that. I'll try that out! Thanks again :+1:

**Comment by chanb** [comment]
Seems like having a smaller actor learning rate already helps. I'll try that adaptive regime if it still fails.

Note: Just for completeness I wasn't updating the agent every timestep because my implementation seems to be fairly inefficient so it takes a long time to run. Not sure if know if the wall time for this algorithm is significantly higher compared to algorithm like SAC because of the extra computations.
